### PR TITLE
Added Multibundle RequireJS example

### PR DESCRIPTION
**Pull request by crwang**
# What/Why

Added an example of how to add requirejs bundles into a Rendr App.

This is a much more optimized way to serve up the javascript code on demand.
# How Tested

Manually and updated the default tests for the examples.
# Other Notes

This is an example, but it has not been optimized yet for developer efficiency (no need to build the bundles every time there's a change in development).  That will be added as a second step.

@alexindigo 

Finally, getting this PR going.  Would you mind helping me review this?

### CODE REVIEW

**Comment by alexindigo** [comment]
@crwang Looks awesome. Thanks a lot. I'm still going through all the changes, but so far I'm in love with `README` :) And `Gruntfile` looks good.

Tagging my partners in (rendr+requirejs) crime, for more fun and better feedback.

/cc @bigethan @jskulski @purusho @abhisheksavant @ingiulio \m/

File: 08_multibundle_requirejs/package.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "mocha --ui bdd --reporter spec  ./test --recursive",
+    "start": "node index.js",
+    "postinstall": "test -f ../../package.json && npm install ../../ || echo",
+    "postupdate": "test -f ../../package.json && npm install ../../ || echo"
+  },
+  "dependencies": {
+    "amdefine": "~0.1.0",
```
Comment (alexindigo) [inline 08_multibundle_requirejs/package.json]: `global-define` has this as it's dependency, so you probably don't need it at this level.

File: 08_multibundle_requirejs/test/app/collections/repos_bundle/repos.test.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+var should = require("should");
+var Repos = require('../../../../app/collections/repos_bundle/repos');
```
Comment (alexindigo) [inline 08_multibundle_requirejs/test/app/collections/repos_bundle/repos.test.js]: I think you can use `global-define` with tests too,
although probably not necessary.

File: 08_multibundle_requirejs/app/lib/handlebarsHelpers.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+if (typeof define !== 'function') {
+  var define = require('amdefine')(module);
+}
```
Comment (bigethan) [inline 08_multibundle_requirejs/app/lib/handlebarsHelpers.js]: Is this bit required here because of handlebars?

File: 08_multibundle_requirejs/Gruntfile.js [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                        'app': 'app',
+                        'assets': 'assets',
+                        // needed for rendr modules
+                        'rendr': 'node_modules/rendr'
+                    },
+                    preserveLicenseComments: false
+                },
+
+                // Creates `<destination>/common.<hash>.js` file that includes all the modules specified in the bundle,
+                // shared modules between all the pages.
```
Comment (bigethan) [inline 08_multibundle_requirejs/Gruntfile.js]: Would it make sense to pull these bits (`common`, `user`,`repo`) out into their own files? I could see the gruntfile getting overly large very quickly.

File: 08_multibundle_requirejs/Gruntfile.js [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                },
+                files: {
+                    'public/js/app/templates/compiledTemplates.js': ['app/templates/**/*.hbs',
+                        '!' + usersBundleTemplatePath, // Exclude since we'll build it separate
+                        '!' + reposBundleTemplatePath, // Exclude since we'll build it separate
+                        '!app/templates/__layout.hbs'
+                    ],
+                    'public/js/app/templates/users_bundle/compiledTemplates.js': usersBundleTemplatePath,
+                    'public/js/app/templates/repos_bundle/compiledTemplates.js': reposBundleTemplatePath
+                }
```
Comment (bigethan) [inline 08_multibundle_requirejs/Gruntfile.js]: this `files` section feels like an odd mix of magic strings and well named variables. No real suggestions (more variables?) just feels off, sorry for not being more constructive.

**Comment by bigethan** [comment]
I like seeing the evolution of this! Using the directory structure to bundle is nice. This feature is a huge performance win for big Rendr apps.  Do other SPA or isomorphic setups do this kinda thing at all?

**Comment by bigethan** [comment]
to be clear, :heart: :heart: :heart: - nice work @crwang and @alexindigo

File: 08_multibundle_requirejs/Gruntfile.js (hunk 4 above)
Comment (alexindigo) [inline 08_multibundle_requirejs/Gruntfile.js]: I think for example apps single Gruntfile is more descriptive and more familiar to novice eye.
When real project grow eve can find their way to make it more manageable. Like ditching grunt and switching to npm scripts :)

File: 08_multibundle_requirejs/Gruntfile.js (hunk 5 above)
Comment (alexindigo) [inline 08_multibundle_requirejs/Gruntfile.js]: This is something we haven't (fully) achieved either, and I "blame" Rendr original flow – having all the templates loaded at all times. What we have in Trulia now is just half-way to the real solution, pre compiled templates required with views. 
And for this example it might require different handlebars adapter. (Need to investigate).

Better way would be to require templates as is using requires plugins (`'hbs!path/to/template'`). This is next step for us in Trulia and probably it will propagate to this (or another) example.

File: 08_multibundle_requirejs/app/lib/handlebarsHelpers.js (hunk 3 above)
Comment (alexindigo) [inline 08_multibundle_requirejs/app/lib/handlebarsHelpers.js]: Good catch, I think we can have global-define to take care of this.

**Comment by alexindigo** [comment]
Thanks for the input @bigethan :) It gives me warm feeling of the good ol' days :)

File: 08_multibundle_requirejs/Gruntfile.js (hunk 5 above)
Comment (alexindigo) [inline 08_multibundle_requirejs/Gruntfile.js]: And it seems like something we would do if we were thinking about breakdown from the very beginning :)

Comment (alexindigo) [inline 08_multibundle_requirejs/Gruntfile.js]: Thinking more over it, I think I got the same feeling, that incorporating bundles into app structure makes it less (like in "no") flexible. And developers need to know what bundles have what modules.

It would definitely give better exposure to bundles themselves and reduce "cross pollination". Which will be good for performance, but maybe not as good codebase evolution.

It feels like old argument Folders vs. Tags in email universe.

As one of the examples where our current structure (bundles specified only in the confit) helped us is when ad blockers were introduced in iOS9 and we had to pull some of the "hot" files into "optional" bundle that we expect not to load in some cases.

Having bundle name as part of the module id (require path) would make it harder and more confusing.

On other hand it might be easier to grasp when you just starting to understand the whole concept.

Bottom line we need more examples. :)
And thanks to @crwang for breaking ground on it. :)

**Comment by alexindigo** [comment]
PS. @bigethan did you see this thing? :) https://github.com/rendrjs/rendr-app-tests

File: 08_multibundle_requirejs/Gruntfile.js (hunk 5 above)
Comment (crwang) [inline 08_multibundle_requirejs/Gruntfile.js]: I agree that this is a bit messy, especially with the `'!' + usersBundleTemplatePath, ` code, but I feel like until we can break out the paths to be more elegant, this way is hopefully good enough for an example.  :smile: 

Once the paths can be something like this: https://github.com/rendrjs/rendr/issues/493, this sort of files setting could be a lot cleaner.

**Comment by crwang** [comment]
I think other SPAs are starting to use webpack which is also very nice.  From my limited research, it seems to be very well done, but at the moment, less customizable than requirejs bundles.

As far as other isomorphic, I personally am not aware of something that is as performant as this approach currently out there.

**Comment by crwang** [comment]
Thanks for the feedback @alexindigo and @bigethan .  

I've updated some based on the feedback given and some formatting issues in the readme.

Alex, if you think it looks ok, it's better if you or @saponifi3d merges it in, philosophically.  Normally, I think that the author shouldn't be the one merging the code, it should be a reviewer.  :smile:

**Comment by alexindigo** [comment]
:+1:

**Comment by alexindigo** [comment]
@crwang Done.
